src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -582,31 +601,82 @@ void QgsSymbolLegendNode::updateLabel()
+    return QString();
 
-    mLabel = mUserLabel.isEmpty() ? layerName : mUserLabel;
-    if ( showFeatureCount && vl && vl->featureCount() >= 0 )
-      mLabel += QStringLiteral( " [%1]" ).arg( vl->featureCount() );
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
+
+  if ( label.isEmpty() )
+  {
+    if ( vl )
+    {
+      context.appendScope( vl->createExpressionContextScope() );
+      const QString symLabel = symbolLabel();
+      if ( ! mLayerNode->labelExpression().isEmpty() )
+        mLabel = evaluateLabelExpression( "[%" + mLayerNode->labelExpression() + "%]", context );
+      else if ( mLabel.contains( "[%" ) )
+        mLabel = evaluateLabelExpression( symLabel, context );
+    }
+    return mLabel;
   }
   else
   {
-    mLabel = mUserLabel.isEmpty() ? mItem.label() : mUserLabel;
-    if ( showFeatureCount && vl )
+    QString eLabel;
+    if ( vl )
     {
-      qlonglong count = vl->featureCount( mItem.ruleKey() );
-      mLabel += QStringLiteral( " [%1]" ).arg( count != -1 ? QLocale().toString( count ) : tr( "N/A" ) );
+      context.appendScope( vl->createExpressionContextScope() );
+      if ( ! mLayerNode->labelExpression().isEmpty() )
+        eLabel = evaluateLabelExpression( label + "[%" + mLayerNode->labelExpression() + "%]", context );
+      else if ( label.contains( "[%" ) )
+        eLabel = evaluateLabelExpression( label, context );
     }
+    return eLabel;
   }
-
-  emit dataChanged();
 }
 
+QgsExpressionContextScope *QgsSymbolLegendNode::createSymbolScope() const
+{
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
+
+  QgsExpressionContextScope *scope = new QgsExpressionContextScope( tr( "Symbol scope" ) );
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_label" ), symbolLabel().remove( "[%" ).remove( "%]" ), true ) );
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_id" ), mItem.ruleKey(), true ) );
+  if ( vl )
+  {
+    vl->countSymbolFeatures();
+    //QVariantList featureIds = QVariantList();
+    //const QgsFeatureIds fids = vl->featureIds( mItem.ruleKey() );
+    //if ( !fids.empty() )
+    //{ featureIds.reserve( fids.count() );
+    //  for ( QgsFeatureId fid : fids )
+    //    featureIds << static_cast<qint64>( fid );
+    //}
+    //scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_feature_ids" ), featureIds, true ) );
+    scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_count" ), QVariant::fromValue( vl->featureCount( mItem.ruleKey() ) ), true ) );
+  }
+  return scope;
+}
 
+QString QgsSymbolLegendNode::evaluateLabelExpression( const QString &label, QgsExpressionContext &context ) const
+{
+  QgsExpressionContextScope *symbolScope = createSymbolScope();
+  context.appendScope( symbolScope );
+  QString eLabel = QgsExpression().replaceExpressionText( label, &context );
```

Comment:
That's a static method I guess?

```suggestion
  QString eLabel = QgsExpression::.replaceExpressionText( label, &context );
```

Reply:
I did not think that such a thing was doable but I'll give it a try.